File: Tps/Tp4/tp04.java
/* 
Separando cada elemento do arquivo csv
AppID: 897820
Name:Reigns: Game of Thrones
Release date: "Oct 18, 2018"
Estimated owners: 75000
Price: 3.99
Supported languages: "['English', 'French', 'German', 'Spanish - Spain', 'Italian', 'Japanese', 'Korean', 'Portuguese - Brazil', 'Russian', 'Simplified Chinese', 'Traditional Chinese']"
Metacritic score: 84
User score: 0
Achievements: 10
Publishers: Devolver Digital
Developers: Nerial
Categories: "Single-player,Steam Achievements,Full controller support"
Genres: "Adventure,Indie,RPG"
Tags: "RPG,Indie,Card Game,Adventure,Choices Matter,2D,Medieval,Singleplayer,Minimalist,Choose Your Own Adventure,Strategy,Story Rich,Dark Humor,Great Soundtrack,Difficult,Dark Comedy,Funny,Comedy,Cute"
*/


import java.io.*;
import java.util.*;

class Game 
{
    private int appId;
    private String name;
    private String releaseDate;
    private int estimatedOwners;
    private double price;
    private String[] supportedLanguages;
    private int metacriticScore;
    private double userScore;
    private int achievements;
    private String[] publishers;
    private String[] developers;
    private String[] categories;
    private String[] genres;
    private String[] tags;

    public int getappId() 
    { 
        return appId; 
    }
    public void setappId(int appId) 
    { 
        this.appId = appId; 
    }

    public String getName() 
    { 
        return name; 
    }
    public void setName(String name) 
    { 
        this.name = name; 
    }

    public String getReleaseDate() 
    { 
        return releaseDate; 
    }
    public void setReleaseDate(String releaseDate) 
    { 
        this.releaseDate = releaseDate; 
    }

    public int getEstimatedOwners() 
    { 
        return estimatedOwners; 
    }
    public void setEstimatedOwners(int estimatedOwners) 
    { 
        this.estimatedOwners = estimatedOwners; 
    }

    public double getPrice() 
    { 
        return price; 
    }
    public void setPrice(double price) 
    { 
        this.price = price; 
    }
    
    public String[] getSupportedLanguages() 
    { 
        return supportedLanguages; 
    }
    public void setSupportedLanguages(String[] supportedLanguages) 
    { 
        this.supportedLanguages = supportedLanguages; 
    }

    public int getMetacriticScore() 
    { 
        return metacriticScore; 
    }
    public void setMetacriticScore(int metacriticScore) 
    { 
        this.metacriticScore = metacriticScore; 
    }

    public double getUserScore() 
    { 
        return userScore; 
    }
    public void setUserScore(double userScore) 
    { 
        this.userScore = userScore; 
    }

    public int getAchievements() 
    { 
        return achievements; 
    }
    public void setAchievements(int achievements) 
    { 
        this.achievements = achievements;
    }

    public String[] getPublishers() 
    { 
        return publishers; 
    }
    public void setPublishers(String[] publishers) 
    { 
        this.publishers = publishers; 
    }

    public String[] getDevelopers() 
    { 
        return developers; 
    }
    public void setDevelopers(String[] developers) 
    { 
        this.developers = developers; 
    }

    public String[] getCategories() 
    { 
        return categories; 
    }
    public void setCategories(String[] categories) 
    { 
        this.categories = categories; 
    }

    public String[] getGenres() 
    { 
        return genres; 
    }
    public void setGenres(String[] genres) 
    { 
        this.genres = genres; 
    }
    
    public String[] getTags() 
    { 
        return tags; 
    }
    public void setTags(String[] tags) 
    { 
        this.tags = tags; 
    }

    public static Game ler(String linha) 
    {
        Game g = new Game();
        String[] campos = splitParaLinguaguem(linha);

        g.appId = Integer.parseInt(colocarEspaco(campos, 0));
        g.name = colocarEspaco(campos, 1);
        g.releaseDate = formatarData(colocarEspaco(campos, 2));
        g.estimatedOwners = Integer.parseInt(colocarEspaco(campos, 3));
        g.price = Double.parseDouble(colocarEspaco(campos, 4));
        g.supportedLanguages = stringParaArray(colocarEspaco(campos, 5));
        g.metacriticScore = Integer.parseInt(colocarEspaco(campos, 6));
        g.userScore = Double.parseDouble(colocarEspaco(campos, 7));
        g.achievements = Integer.parseInt(colocarEspaco(campos, 8));
        g.publishers = stringParaArray(colocarEspaco(campos, 9));
        g.developers = stringParaArray(colocarEspaco(campos, 10));
        g.categories = stringParaArray(colocarEspaco(campos, 11));
        g.genres = stringParaArray(colocarEspaco(campos, 12));
        g.tags = stringParaArray(colocarEspaco(campos, 13));

        return g;
    }

    private static String[] splitParaLinguaguem(String linha) 
    {
        String[] campos = new String[100];
        int count = 0;

        StringBuilder campoAtual = new StringBuilder();
        boolean dentroAspas = false;
        int colchetes = 0;

        for (int i = 0; i < linha.length(); i++) 
        {
            char c = linha.charAt(i);

            if (c == '\"') {
                dentroAspas = !dentroAspas; 
                campoAtual.append(c);
            } else if (c == '[') {
                colchetes++;
                campoAtual.append(c);
            } else if (c == ']') {
                colchetes--;
                campoAtual.append(c);
            } else if (c == ',' && !dentroAspas && colchetes == 0) {
                campos[count++] = campoAtual.toString().trim();
                campoAtual.setLength(0);
            } else {
                campoAtual.append(c);
            }
        }

        // Adiciona o último campo
        campos[count++] = campoAtual.toString().trim();

        // Cria um array do tamanho exato
        String[] resultado = new String[count];
        for (int i = 0; i < count; i++) {
            resultado[i] = campos[i];
        }

        return resultado;
    }


    private static String formatarData(String data) 
    {
        if (data == null || data.isEmpty()) 
        {
            return "";
        }

        String[] partes = data.split(" ");
        
        if (partes.length < 3) 
        {
            return data;
        }

        String mesTexto = partes[0];
        String diaTexto = partes[1].replace(",", "");
        String anoTexto = partes[2];

        int dia = Integer.parseInt(diaTexto);

        String diaFormatado;
        if (dia < 10) 
        {
            diaFormatado = "0" + dia;
        } 
        else 
        {
            diaFormatado = String.valueOf(dia);
        }

        String mesFormatado;
        switch (mesTexto) 
        {
            case "Jan": mesFormatado = "01"; break;
            case "Feb": mesFormatado = "02"; break;
            case "Mar": mesFormatado = "03"; break;
            case "Apr": mesFormatado = "04"; break;
            case "May": mesFormatado = "05"; break;
            case "Jun": mesFormatado = "06"; break;
            case "Jul": mesFormatado = "07"; break;
            case "Aug": mesFormatado = "08"; break;
            case "Sep": mesFormatado = "09"; break;
            case "Oct": mesFormatado = "10"; break;
            case "Nov": mesFormatado = "11"; break;
            case "Dec": mesFormatado = "12"; break;
            default: mesFormatado = "00"; break;
        }

        return diaFormatado + "/" + mesFormatado + "/" + anoTexto;
    }

    private static String colocarEspaco(String[] c, int i) 
    {
        if (i >= c.length) return "";
        return c[i].replace("\"", "").trim();
    }

    private static String[] stringParaArray(String campo) 
    {
        campo = campo.replace("[", "").replace("]", "").replace("'", "").replace("\"", "").trim();
        if (campo.isEmpty()) return new String[0];
        return campo.split(",");
    }

    public String formatarSaida() 
    {
        return "=> " + appId + " ## " + name + " ## " + releaseDate + " ## " + estimatedOwners
                + " ## " + price + " ## " + Arrays.toString(supportedLanguages) + " ## " + metacriticScore
                + " ## " + userScore + " ## " + achievements + " ## " + Arrays.toString(publishers)
                + " ## " + Arrays.toString(developers) + " ## " + Arrays.toString(categories)
                + " ## " + Arrays.toString(genres) + " ## " + Arrays.toString(tags) + " ##";
    }
}

public class tp04
{
    public static void main(String[] args) 
    {
        Game[] lista = new Game[100000];
        int count = 0;

        File file = new File("games.csv");
        Scanner arquivo = null;

        try 
        {
            arquivo = new Scanner(file);
        } 
        catch (FileNotFoundException e) 
        {
            System.out.println(""); //arquivo nao encontrado
            return;
        }

        if (arquivo.hasNextLine()) //pular a primeira linha
        { 
            arquivo.nextLine();
        }   
        
        while (arquivo.hasNextLine() && count < lista.length) 
        {
            String linha = arquivo.nextLine();
            if (linha.trim().isEmpty()) continue;
            lista[count++] = Game.ler(linha);
        }
        arquivo.close();

        Scanner sc = new Scanner(System.in); //entrada do id

        while (true) 
        {
            String entrada = sc.nextLine().trim();

            if (entrada.equals("FIM")) 
            {
                break;
            }

            // Verifica se a entrada contém apenas números
            boolean entradaValida = true;
            for (int i = 0; i < entrada.length(); i++) 
            {
                char c = entrada.charAt(i);
                if (c < '0' || c > '9') 
                {
                    entradaValida = false;
                    break;
                }
            }

            if (entradaValida) 
            {
                int id = Integer.parseInt(entrada);
                Game jogoEncontrado = null;

                // Busca o jogo pelo ID
                for (int i = 0; i < count; i++) 
                {
                    if (lista[i].getappId() == id) 
                    {
                        jogoEncontrado = lista[i];
                        break;
                    }
                }

                if (jogoEncontrado != null)
                {
                    System.out.println(jogoEncontrado.formatarSaida());
                } 
                else 
                {
                    System.out.println(""); // não achou
                }

            } 
            else 
            {
                System.out.println(""); // entrada inválida
            }


        }

        sc.close();
    }
}

File: Tps/Tp4/tp04.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define MAX_STRING 500
#define MAX_ARRAY 100
#define MAX_LINE 5000

typedef struct 
{
    int appId;
    char name[MAX_STRING];
    char releaseDate[MAX_STRING];
    int estimatedOwners;
    double price;
    char supportedLanguages[MAX_ARRAY][MAX_STRING];
    int supportedLanguagesCount;
    int metacriticScore;
    double userScore;
    int achievements;
    char publishers[MAX_ARRAY][MAX_STRING];
    int publishersCount;
    char developers[MAX_ARRAY][MAX_STRING];
    int developersCount;
    char categories[MAX_ARRAY][MAX_STRING];
    int categoriesCount;
    char genres[MAX_ARRAY][MAX_STRING];
    int genresCount;
    char tags[MAX_ARRAY][MAX_STRING];
    int tagsCount;
} Game;

void initializeGame(Game *game) 
{
    game->appId = 0;
    strcpy(game->name, "");
    strcpy(game->releaseDate, "");
    game->estimatedOwners = 0;
    game->price = 0.0;
    game->supportedLanguagesCount = 0;
    game->metacriticScore = -1;
    game->userScore = -1.0;
    game->achievements = 0;
    game->publishersCount = 0;
    game->developersCount = 0;
    game->categoriesCount = 0;
    game->genresCount = 0;
    game->tagsCount = 0;
}

void removeAspas(char *str) 
{
    int j = 0;
    for (int i = 0; str[i] != '\0'; i++)
        if (str[i] != '"' && str[i] != '\'')
            str[j++] = str[i];
    str[j] = '\0';
}

bool isSpace(char c) 
{
    return (c == ' ' || c == '\t' || c == '\n' || c == '\r');
}

void trim(char *str) 
{
    int start = 0;
    while (isSpace(str[start])) start++;
    int end = strlen(str) - 1;
    while (end >= start && isSpace(str[end])) end--;

    int j = 0;
    for (int i = start; i <= end; i++) 
    {
        str[j++] = str[i];
    }
    str[j] = '\0';
}

bool isDigit(char c) 
{ 
    return (c >= '0' && c <= '9'); 
}

int stringParaInt(char *str) 
{
    int res = 0;
    int i = 0;
    while (str[i] != '\0') 
    {
        if (isDigit(str[i])) 
        {
            res = res * 10 + (str[i] - '0');
        }
        i++;
    }
    return res;
}

double stringParaDouble(char *str) 
{
    double res = 0.0;
    int i = 0;
    while (str[i] != '\0' && str[i] != '.') 
    {
        if (isDigit(str[i])) res = res * 10 + (str[i] - '0');
        i++;
    }
    if (str[i] == '.') 
    {
        i++;
        double divisor = 10.0;
        while (str[i] != '\0') 
        {
            if (isDigit(str[i])) 
            {
                res += (str[i] - '0') / divisor;
                divisor *= 10;
            }
            i++;
        }
    }
    return res;
}

int separarCamposCSV(char *linha, char campos[][MAX_STRING]) 
{
    int count = 0;
    bool dentroAspas = false;
    char campoAtual[MAX_STRING] = "";
    int pos = 0;

    for (int i = 0; linha[i] != '\0'; i++) 
    {
        char c = linha[i];
        if (c == '"') dentroAspas = !dentroAspas;
        else if (c == ',' && !dentroAspas) 
        {
            campoAtual[pos] = '\0';
            trim(campoAtual);
            strncpy(campos[count++], campoAtual, MAX_STRING - 1);
            pos = 0;
            campoAtual[0] = '\0';
        } else campoAtual[pos++] = c;
    }
    if (pos > 0) 
    {
        campoAtual[pos] = '\0';
        trim(campoAtual);
        strncpy(campos[count++], campoAtual, MAX_STRING - 1);
    }
    return count;
}

void processarArray(char *campo, char array[][MAX_STRING], int *count) 
{
    *count = 0;
    if (strlen(campo) == 0) return;

    char temp[MAX_STRING];
    int len = strlen(campo);
    int start = 0, end = len - 1;

    if (campo[0] == '[') start++;
    if (campo[len - 1] == ']') end--;

    int j = 0;
    for (int i = start; i <= end; i++) temp[j++] = campo[i];
    temp[j] = '\0';

    bool dentroAspas = false;
    char itemAtual[MAX_STRING] = "";
    int pos = 0;
    for (int i = 0; temp[i] != '\0'; i++) 
    {
        char c = temp[i];
        if (c == '"') dentroAspas = !dentroAspas;
        else if (c == ',' && !dentroAspas) 
        {
            itemAtual[pos] = '\0';
            trim(itemAtual);
            removeAspas(itemAtual);
            if (strlen(itemAtual) > 0)
                strncpy(array[(*count)++], itemAtual, MAX_STRING - 1);
            pos = 0;
            itemAtual[0] = '\0';
        } 
        else itemAtual[pos++] = c;
    }
    if (pos > 0) 
    {
        itemAtual[pos] = '\0';
        trim(itemAtual);
        removeAspas(itemAtual);
        if (strlen(itemAtual) > 0)
            strncpy(array[(*count)++], itemAtual, MAX_STRING - 1);
    }
}

void formatarData(char *data) 
{
    if (strlen(data) == 0) return;

    char mes[4] = {0}, dia[3] = {0}, ano[5] = {0};
    int i = 0, j = 0;

    for (i = 0; i < 3 && data[i]; i++) 
    {
        mes[i] = data[i];
    }
    mes[3] = '\0';

    while (data[i] && data[i] == ' ') i++;

    j = 0;
    while (data[i] && data[i] != ',' && j < 2) 
    {
        if (isDigit(data[i])) 
        {
            dia[j++] = data[i];
        }
        i++;
    }
    dia[j] = '\0';
    int diaInt = stringParaInt(dia);

    while (data[i] && (data[i] == ',' || data[i] == ' ')) i++;

    j = 0;
    while (data[i] && j < 4) 
    {
        ano[j++] = data[i++];
    }
    ano[j] = '\0';

    char *mesNum = "01";
    if (strcmp(mes, "Jan") == 0) mesNum = "01";
    else if (strcmp(mes, "Feb") == 0) mesNum = "02";
    else if (strcmp(mes, "Mar") == 0) mesNum = "03";
    else if (strcmp(mes, "Apr") == 0) mesNum = "04";
    else if (strcmp(mes, "May") == 0) mesNum = "05";
    else if (strcmp(mes, "Jun") == 0) mesNum = "06";
    else if (strcmp(mes, "Jul") == 0) mesNum = "07";
    else if (strcmp(mes, "Aug") == 0) mesNum = "08";
    else if (strcmp(mes, "Sep") == 0) mesNum = "09";
    else if (strcmp(mes, "Oct") == 0) mesNum = "10";
    else if (strcmp(mes, "Nov") == 0) mesNum = "11";
    else if (strcmp(mes, "Dec") == 0) mesNum = "12";

    char novaData[11] = {0};
    if (diaInt < 10) 
    {
        novaData[0] = '0';
        novaData[1] = diaInt + '0';
    } 
    else 
    {
        novaData[0] = (diaInt / 10) + '0';
        novaData[1] = (diaInt % 10) + '0';
    }
    novaData[2] = '/';
    novaData[3] = mesNum[0];
    novaData[4] = mesNum[1];
    novaData[5] = '/';
    novaData[6] = ano[0];
    novaData[7] = ano[1];
    novaData[8] = ano[2];
    novaData[9] = ano[3];
    novaData[10] = '\0';

    strcpy(data, novaData);
}

int parseEstimatedOwners(char *str)
{
    char temp[MAX_STRING];
    int j = 0;
    for (int i = 0; str[i] != '\0'; i++)
        if (isDigit(str[i])) temp[j++] = str[i];
    temp[j] = '\0';
    return stringParaInt(temp);
}

void lerGame(Game *game, char *linha) 
{
    initializeGame(game);
    char campos[30][MAX_STRING];
    int numCampos = separarCamposCSV(linha, campos);
    if (numCampos > 0) game->appId = stringParaInt(campos[0]);
    if (numCampos > 1) strncpy(game->name, campos[1], MAX_STRING - 1);
    if (numCampos > 2) 
    {
        strncpy(game->releaseDate, campos[2], MAX_STRING - 1);
        formatarData(game->releaseDate);
    }
    if (numCampos > 3) game->estimatedOwners = parseEstimatedOwners(campos[3]);
    if (numCampos > 4) 
    {
        if (strcmp(campos[4], "Free to Play") == 0)
            game->price = 0.0;
        else
            game->price = stringParaDouble(campos[4]);
    }
    if (numCampos > 5) processarArray(campos[5], game->supportedLanguages, &game->supportedLanguagesCount);
    if (numCampos > 6) game->metacriticScore = stringParaInt(campos[6]);
    if (numCampos > 7 && strcmp(campos[7], "tbd") != 0) game->userScore = stringParaDouble(campos[7]);
    if (numCampos > 8) game->achievements = stringParaInt(campos[8]);
    if (numCampos > 9) processarArray(campos[9], game->publishers, &game->publishersCount);
    if (numCampos > 10) processarArray(campos[10], game->developers, &game->developersCount);
    if (numCampos > 11) processarArray(campos[11], game->categories, &game->categoriesCount);
    if (numCampos > 12) processarArray(campos[12], game->genres, &game->genresCount);
    if (numCampos > 13) processarArray(campos[13], game->tags, &game->tagsCount);
}

void formatarArrayParaSaida(char array[][MAX_STRING], int count, char *resultado) 
{
    strcpy(resultado, "[");
    for (int i = 0; i < count; i++) 
    {
        strcat(resultado, array[i]);
        if (i < count - 1) strcat(resultado, ", ");
    }
    strcat(resultado, "]");
}

void formatarSaida(Game *game) 
{
    char languages[1000], pubs[1000], devs[1000], cats[1000], gens[1000], tags[1000];
    formatarArrayParaSaida(game->supportedLanguages, game->supportedLanguagesCount, languages);
    formatarArrayParaSaida(game->publishers, game->publishersCount, pubs);
    formatarArrayParaSaida(game->developers, game->developersCount, devs);
    formatarArrayParaSaida(game->categories, game->categoriesCount, cats);
    formatarArrayParaSaida(game->genres, game->genresCount, gens);
    formatarArrayParaSaida(game->tags, game->tagsCount, tags);

    printf("=> %d ## %s ## %s ## %d ## %.2f ## %s ## %d ## %.1f ## %d ## %s ## %s ## %s ## %s ## %s ##\n",
        game->appId,
        game->name, game->releaseDate, game->estimatedOwners, game->price,
        languages, game->metacriticScore, game->userScore, game->achievements,
        pubs, devs, cats, gens, tags);
}

bool isInteger(char *str) 
{
    if (str == NULL || strlen(str) == 0) return false;
    for (int i = 0; str[i] != '\0'; i++)
        if (!isDigit(str[i])) return false;
    return true;
}

int main() 
{
    FILE *arquivo = fopen("/tmp/games.csv", "r");
    if (arquivo == NULL) 
    {
        printf("Erro ao abrir o arquivo\n");
        return 1;
    }

    char entrada[100];
    char linha[MAX_LINE];
    fgets(linha, MAX_LINE, arquivo);

    while (1) 
    {
        if (fgets(entrada, sizeof(entrada), stdin) == NULL) break;
        entrada[strcspn(entrada, "\n")] = '\0';
        if (strcmp(entrada, "FIM") == 0) break;

        if (isInteger(entrada)) 
        {
            int idProcurado = stringParaInt(entrada);
            bool encontrado = false;
            rewind(arquivo);
            fgets(linha, MAX_LINE, arquivo);

            while (fgets(linha, MAX_LINE, arquivo) != NULL) 
            {
                Game g;
                lerGame(&g, linha);
                if (g.appId == idProcurado) 
                {
                    formatarSaida(&g);
                    encontrado = true;
                    break;
                }
            }
            if (!encontrado) printf("\n");
        }
    }

    fclose(arquivo);
    return 0;
}